Add minigo

yhliang2018 opened the pull request:
The original MiniGo can be roughly divided into three main parts: DualNet model, MCTS, and Go-related domain knowledge. This PR is mainly about the DualNet model part, and asks for reviews on the following four files:
minigo.py: The main entry to run minigo.
dualnet_model.py: The model definition.
dual_net.py: The run loop of DualNet.
model_utils: Utils for DualNet model, including model parameters and other misc.

For other py files, feel free to add any comments if you have. I am still working on them, mainly about code clean up and Python2 compatibility.

karmel commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+        selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
+                 params)
+      games = tf.gfile.Glob(os.path.join(selfplay_dir, model_name, '*.zz'))
+      if len(games) >= params.max_games_per_generation:
```
I fear while True loops. Why is this necessary? When is the number of games less than one? When is the number of games > 1? Would this work if we did `for _ in params.max_games_per_generation` and then break early if the num games grows faster than one per generation?

karmel commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+      help='A boolean to explicitly generate holdout data for validation.')
+
+  FLAGS, unparsed = parser.parse_known_args()
+  tf.app.run(main=main, argv=[sys.argv[0]] + unparsed)
```
We will have to change this when going into official, though doesn't necessarily need to happen immediately.

karmel commented on research/minigo/model_utils.py:
```diff
@@ -0,0 +1,206 @@
+# ==============================================================================
+"""Utilities for DualNet model."""
+
+# pylint: disable=g-bad-import-order
```
Ditto on disabling.

karmel commented on research/minigo/model_utils.py:
```diff
@@ -0,0 +1,206 @@
+
+from tensorflow.python.training.summary_io import SummaryWriterCache
+
+# Regular expression of model number and name.
```
Comments! What format is this capturing? Remember, one of my goals is to never have to manually parse regex strings.

karmel commented on research/minigo/model_utils.py:
```diff
@@ -0,0 +1,206 @@
+
+def compute_update_ratio(weight_tensors, before_weights, after_weights):
+  """Compute the ratio of gradient norm to weight norm."""
+  deltas = [after - before for after,
```
Args, returns-- are these tensors, arrays? What is this computing, and why?

karmel commented on research/minigo/model_utils.py:
```diff
@@ -0,0 +1,206 @@
+  def after_run(self, run_context, run_values):
+    """Called after each call to run()."""
+    if self.before_weights is not None:
+      after_weights = run_context.session.run(self.weight_tensors)
```
Why do we call session run on these instead of just returning them in run_values? Wouldn't it be more straightforward to make sure they are returned out of the step in question, rather than grabbing them here?

karmel commented on research/minigo/model_utils.py:
```diff
@@ -0,0 +1,206 @@
+
+  # dual_net
+  # How many positions to look at per generation.
+  # Per AGZ, 2048 minibatch * 1k = 2M positions/generation
```
It took me a while to figure out what AGZ is; can you spell out?

yhliang2018 replied:
Will do.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+import symmetries
+
+
+class DualNetwork(object):
```
Calling this the Network seems unclear, because that implies it is the actual layers. DualNetRunner or something?

yhliang2018 replied:
Sure. Will rename it.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+      params: Hyper parameters for the dual network
```
hyperparameters; please provide expected type

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    self.inference_input = None
+    self.inference_output = None
+    config = tf.ConfigProto()
+    config.gpu_options.allow_growth = True
```
This is interesting. Why do we set this here? That changes the behavior WRT GPU mem usage. @bitfort -- what does the addition/removal of this change? My understanding is that this tries to limit gpu mem, but has some strange effects in certain cases.

yhliang2018 replied:
I checked this link for using GPUs: https://www.tensorflow.org/programmers_guide/using_gpu. From this post, I think to set the allow_growth to True is trying to allow the GPU memory growth during the process execution, rather than pre-mapping all GPU memory to all GPUs at the begging. Not sure if this is correct. @bitfort Could you clarify on this?

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    """Initialize the dual network from saved model/checkpoints.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
```
Save file is used to initialize weights, which is unclear from this, and apparently takes a value of None? Can you make the docstring more explicit here?

yhliang2018 replied:
Save file here will be feed with a saved checkpoint file, which is not None. I will put an example in the docstring.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    tf.train.Saver().restore(self.sess, save_file)
+
+  def run(self, position, use_random_symmetry=True):
+    """Compute the policy and value output for a given position."""
```
Comments-- what is symmetry here, what is the position, what is returned.

yhliang2018 replied:
Have updated the docstring.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    return probs[0], values[0]
+
+  def run_many(self, positions, use_random_symmetry=True):
+    processed = list(map(features.extract_features, positions))
```
Ditto the above on docstrings. Also, comment on what the expected values are here-- what features are extracted from the positions, what is the expected format, how are they used?

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    if use_random_symmetry:
+      syms_used, processed = symmetries.randomize_symmetries_feat(processed)
+    outputs = self.sess.run(
+        self.inference_output, feed_dict={self.inference_input: processed})
```
Most people don't know what feed_dict is, and how this call works. Please add comments describing what is happening here.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+  labels = {
+      'pi_tensor': tf.placeholder(tf.float32, [None, go.N * go.N + 1]),
+      'value_tensor': tf.placeholder(tf.float32, [None])
+  }
```
Please comment-- what do each of these represent, what will happen when they get fed into the session?

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+
+  This class can restore the model from saved files, and provide inference for
+  given examples.
+  """
```
Why does this exist? It seems to be used for the match, but why is estimator insufficient there? Please comment as to how this differs from the code below for bootstrap, eval, etc.

yhliang2018 replied:
To restore a saved model, estimator needs to restore both the graph definition and the variables, right? I thin here DualNet tries to isolate graph initialization and weights initialization as two models in a play match share the same graph model but with different weights, thus is more efficient.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+  estimator_initial_checkpoint_name = 'model.ckpt-1'
+  save_file = os.path.join(working_dir,
+                           estimator_initial_checkpoint_name)
+  sess = tf.Session(graph=tf.Graph())
```
I don't think you need to pass in tf.Graph here if no config is included.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+  save_file = os.path.join(working_dir,
+                           estimator_initial_checkpoint_name)
+  sess = tf.Session(graph=tf.Graph())
+  with sess.graph.as_default():
```
Why do we do this manually rather than actually creating and running an estimator?

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    model_path: Either a local path or a gs:// path to export model to.
+  """
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params='ignored')
```
You should be able to avoid creating a fake estimator-- tf.train.latest_checkpoint(working_dir) is the same.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+    generation_num: The generation to be trained.
+    params: hyperparams of the model.
+  """
+  assert generation_num > 0, 'Model 0 is random weights'
```
Better to raise a ValueError rather than assert here, since it is the error of the caller.

karmel commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params=params)
+  if checkpoint_name is None:
+    checkpoint_name = estimator.latest_checkpoint()
```
This doesn't seem to be used? (Or else I'm seeing things after 2+ hours of code review :) )

yhliang2018 replied:
You are right, it always evaluate the latest checkpoint in current implementation. The original code can provide validation to any checkpoint, here I just simplify it to the latest one. Will remove it. Thanks!

karmel commented on research/minigo/dualnet_model.py:
```diff
@@ -0,0 +1,236 @@
+
+  Args:
+    inputs: A tensor representing a batch of input features with shape
+      [BATCH_SIZE, go.N, go.N, features.NEW_FEATURES_PLANES].
```
Don't forget to update the docstrings when we change go.N to board_size

karmel commented on research/minigo/dualnet_model.py:
```diff
@@ -0,0 +1,236 @@
+
+Both the policy and value networks share ResNet as a majority of their
+architecture. The neural network consists of a residual tower (ResNet) and two
+separate "heads" for computing the policy and value respectively.
```
Here or in readme it will be important to explain what these different heads are, and what they each output.

karmel commented on research/minigo/dualnet_model.py:
```diff
@@ -0,0 +1,236 @@
+
+  Args:
+    features: tensor with shape
+      [BATCH_SIZE, self.go_size, self.go_size, features.NEW_FEATURES_PLANES]
```
nit: we seem to use go.N, go_size, and self.go_size interchangeably here. We should pick a name and stick with it, esp for this case, where there is no self for the model_fn.

karmel commented on research/minigo/dualnet_model.py:
```diff
@@ -0,0 +1,236 @@
+      'pi_tensor': [BATCH_SIZE, self.go_size * self.go_size + 1]
+      'value_tensor': [BATCH_SIZE]
+    mode: a tf.estimator.ModeKeys (batchnorm params update for TRAIN only)
+    params: a dict of hyperparams
```
nit: params seems to be an object, not a dict

karmel commented on research/minigo/dualnet_model.py:
```diff
@@ -0,0 +1,236 @@
+  # For logging purpose
+  policy_entropy = -tf.reduce_mean(tf.reduce_sum(
+      policy_output * tf.log(policy_output), axis=1))
+
```
Throughout this fn, it would be helpful to have "walk through" explanations for those who are not familiar-- what are each of the losses used for? Why do we care about policy entropy? Etc.

yhliang2018 commented on research/minigo/features.py:
```diff
@@ -0,0 +1,160 @@
+    return onehot_features
+
+
+def planes(num_planes):
```
It tries to indicate the number of planes in the features. For example, for a 19x19 go board, the input feature will be in the shape of [19, 19, 16], where the third dimension is the num_planes. But it's confusing as this feature py file includes both the new feature used in AlphaGo Zero model and the old features in the original AlphaGo. Will remove the old ones to make it easy to follow.

karmel replied:
Also, add this text as a comment.

yhliang2018 commented on research/minigo/mcts.py:
```diff
@@ -0,0 +1,298 @@
+
+All terminology here (Q, U, N, p_UCT) uses the same notation as in the
+AlphaGo (AG) paper.
+"""
```
Sure, will check the AG paper and add it. :)

yhliang2018 commented on research/minigo/mcts.py:
```diff
@@ -0,0 +1,298 @@
+# Dirichlet noise, as a function of go.N
+
+
+def D_NOISE_ALPHA(): return 0.03 * 361 / (go.N ** 2)
```
Have updated it as a function which relies on go board size.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  Args:
+    directory: A given directory
+  """
+  if directory.startswith('gs://'):  # for cloud logging
```
This "gs://" allows user to store all Go related data/model/log into gs bucket, as the resulted data may be very large in terms of size. Of course, local directory always works. Should we remove this feature? @karmel

karmel replied:
Let's remove for now unless we are actively using and testing it. We will likely eventually need to support GCS directly, but we'll cross that bridge when we come to it.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  """Initialize the model with random weights.
+
+  Args:
+    estimator_working_dir: tf.estimator working directory.
```
Yes, it is the model directory the estimator is working on. Will rename it.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+
+  Args:
+    estimator_working_dir: tf.estimator working directory.
+    models_dir: Where to export the first bootstrapped generation.
```
Hmm, sorry for the confusion. The working dir is the directory used by the current model/estimator for output, while the models_dir is a directory to store all the previous trained models. Basically, models_dir stores the last checkpoints of all previous models.

karmel replied:
maybe est_model_dir and trained_models_dir or something to that effect?

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  Args:
+    model_name: The name of the model used for selfplay.
+    models_dir: The path to the network model files.
+    selfplay_dir: Where to write the games'='data/selfplay/.
```
'='  means after the 'base_dir' that is from the user's input. It's from the original codes. Will update it to make it clear.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+    models_dir: The path to the network model files.
+    selfplay_dir: Where to write the games'='data/selfplay/.
+    holdout_dir: Where to write the hold out data'='data/holdout/.
+    sgf_dir: Where to write the sgfs"="sgf/.
```
Will do.

yhliang2018 commented on research/minigo/model_utils.py:
```diff
@@ -0,0 +1,206 @@
+
+
+class UpdateRatioSessionHook(tf.train.SessionRunHook):
+  """Update the ratio of gradient norm to weight norm every N steps."""
```
This is from the original implementation. I think it's safe to remove it if we don't care about it in mlbenchmark. What do you think? @karmel

karmel replied:
Let's remove it then-- thanks.

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  dual_net.export_model(estimator_working_dir, bootstrap_model_path)
+
+
+def selfplay(model_name, models_dir, selfplay_dir, holdout_dir, sgf_dir,
```
Can we pass in the network itself here, instead of the model_name? There is significant overhead (~33%) if we load the network for every game versus caching it.

yhliang2018 replied:
Do you mean we should load the network into memory in the main entry, and pass the network as a parameter to selfplay or other sub-functions that take the network as an input? This is the same like 'network caching', right?

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+
+  with utils.logged_timer('Playing game'):
+    player = selfplay_mcts.play(
+        network, params.readouts, params.resign_threshold, params.verbose)
```
Should this be selfplay readouts? As opposed evaluation of other readouts.

yhliang2018 replied:
Already rename it. Thanks!

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+    The model name of the winner.
+  """
+
+  black_model = os.path.join(models_dir, black_model_name)
```
Again, caching the models is very useful. We should either allow the models to be passed in, or we could add a #games parameters and loop here.

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  evaluate_dir = os.path.join(FLAGS.base_dir, 'sgf/evaluate/')
+
+  print('Creating random initial weights...')
+  bootstrap(estimator_working_dir, models_dir, params)
```
This should be done conditionally... so we can restart a stopped run.

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+    preprocessing.SHUFFLE_BUFFER_SIZE = 1000
+
+  # Set directories for models and datasets
+  if os.path.isdir(FLAGS.base_dir):  # if it already exists, delete it.
```
We should let the directory stay there so we can restart a run which was stopped.

yhliang2018 replied:
Yes, it makes sense. Have updated the code to make restart work.

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  # Start from the bootstrap model to play games.
+  print('Playing some games...')
+  model_name = '000000-bootstrap'
+  for rl_iter in range(params.max_iters_per_pipeline):
```
We should have options to only do the selfplay, or only do the train-eval... this will allow us to achieve process level paralellization easily.

bitfort commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+
+  # How many positions can fit on a graphics card. 256 for 9s, 16 or 32 for 19s.
+  if FLAGS.go_size == 9:
+    params.batch_size = 256
```
We should be able to set batch_size via command line, I think... because graphics cards can very greatly.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  with utils.logged_timer('Building lists of holdout files'):
+    for record_dir in holdout_dirs:
+      if os.path.exists(record_dir):  # make sure holdout dir exists
+        tf_records.extend(tf.gfile.Glob(os.path.join(record_dir, '*.zz')))
```
They acts the same here. Will update it as tfrecord.zz for consistency.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+      print('Validating on the holdout game data...')
+      validate(models_dir, holdout_dir, estimator_working_dir, params)
+
+    _, latest_model_name = model_utils.get_latest_model(models_dir)
```
After the training step, the trained model will be exported to the models_dir. So here the latest_model is the one we are currently training. model_name here represents the best_model_so_far. Here we try to evaluate with the latest model we just got and the best model so far. Will rename it for clarity.

yhliang2018 commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+
+  def input_fn():
+    return preprocessing.get_input_tensors(
+        params.batch_size, tf_records, shuffle_buffer_size=1000,
```
I have added it to the parameters class.

yhliang2018 commented on research/minigo/dualnet_model.py:
```diff
@@ -0,0 +1,236 @@
+
+import tensorflow as tf
+
+_BATCH_NORM_DECAY = 0.997
```
Sure. I can add it into the param class. Do we need to parameterize it when doing benchmark? As it's also a constant value in the ResNet model.

yhliang2018 commented on research/minigo/evaluation.py:
```diff
@@ -0,0 +1,107 @@
+SIMULTANEOUS_LEAVES = 8
+
+
+def play_match(black_net, white_net, games, readouts, sgf_dir, verbosity):
```
This is used in evaluation with two networks, if you check the evaluate step in the minigo.py.

yhliang2018 commented on research/minigo/mcts.py:
```diff
@@ -0,0 +1,298 @@
+import coords
+import go
+
+MAX_DEPTH = (go.N ** 2) * 1.4  # 505 moves for 19x19, 113 for 9x9
```
Yes, already done.

yhliang2018 commented on research/minigo/dual_net.py:
```diff
@@ -0,0 +1,209 @@
+      input_fn, hooks=[update_ratio_hook, tf.train.ProfilerHook(save_secs=600)],
+      max_steps=max_steps)
+
+
```
Th following validate is the actual 'evaluate' step. I guess here to use 'validate' is trying to distinguish from the evaluation with two Go models.

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  gather: Group games played with the same model into larger files of tfexamples
+  train: Train a new model with the selfplay results from the most recent
+    N generations.
+After training, validation can be preformed on the holdout data. Besides, given
```
Make senses, thanks! Need more professional training on technical docs :)

yhliang2018 commented on research/minigo/minigo.py:
```diff
@@ -0,0 +1,406 @@
+  model_path = os.path.join(models_dir, model_name)
+  output_dir = os.path.join(selfplay_dir, model_name)
+  holdout_dir = os.path.join(holdout_dir, model_name)
+  clean_sgf = os.path.join(sgf_dir, model_name, 'clean')
```
Clean sgf has no 'comments' field when writing the file; while full has. Docstring updated.

yhliang2018 commented on research/minigo/preprocessing.py:
```diff
@@ -0,0 +1,218 @@
+# where it started; this and the interleave parameters in preprocessing can give
+# us an approximation of a uniform sampling.  The default of 4M is used in
+# training, but smaller numbers can be used for aggregation or validation.
+SHUFFLE_BUFFER_SIZE = int(2*1e6)
```
Yeah, already done.

yhliang2018 commented on research/minigo/selfplay_mcts.py:
```diff
@@ -0,0 +1,91 @@
+SIMULTANEOUS_LEAVES = 8
+
+
+def play(network, readouts, resign_threshold, verbosity=0):
```
This is the actual network

karmel commented on research/minigo/evaluation.py:
```diff
@@ -0,0 +1,107 @@
+                                   timeper,
+                                   dur))
+
+  return 'B' if (black_win_counts - white_win_counts) > 0.55 * games else 'W'
```
Why 55% and not 50?

yhliang2018 replied:
It's from the original paper. Have updated the comments in docstring.

karmel commented on research/minigo/evaluation.py:
```diff
@@ -0,0 +1,107 @@
+          _file.write(sgfstr)
+        print("Finished game", i, active.result_string)
+        if active.result_string is not None:
+          if active.result_string[0] is 'B':
```
'B' and 'W' should be included as constants in a module somewhere, such that you can reference go.BLACK_NAME and go.WHITE_NAME instead of having the strings typed out in all of the locations. (This can be an actual constant, not a parameter :) )

yhliang2018 replied:
Done.

karmel commented on research/minigo/go.py:
```diff
@@ -0,0 +1,530 @@
+
+
+def is_koish(board, c):
+    'Check if c is surrounded on all sides by 1 color, and return that color'
```
nit: docstring formatting

karmel commented on research/minigo/go.py:
```diff
@@ -0,0 +1,530 @@
+
+
+def is_eyeish(board, c):
+    'Check if c is an eye, for the purpose of restricting MC rollouts.'
```
ditto

karmel commented on research/minigo/gtp_extensions.py:
```diff
@@ -0,0 +1,170 @@
+            raise
+
+# Should this class blatantly reach into the game_obj and frob its tree?  Sure!
+# What are private members?  Python lets you do *anything!*
```
What is this comment referring to?

yhliang2018 replied:
Not relevant and already delete it.

karmel commented on research/minigo/utils.py:
```diff
@@ -0,0 +1,95 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
```
nit: isn't there normally a `#=====` for all of these licenses? Can we add that back?

karmel commented on research/minigo/utils.py:
```diff
@@ -0,0 +1,95 @@
+import re
+import time
+import math
+
```
import order

karmel commented on research/minigo/symmetries.py:
```diff
@@ -0,0 +1,82 @@
+
+def invert_symmetries_pi(symmetries, pis):
+    return [apply_symmetry_pi(invert_symmetry(s), pi)
+            for s, pi in zip(symmetries, pis)]
```
Do we have tests for all of these functions?

yhliang2018 commented:
Thanks a lot, Karmel, Kathy and Victor, for your detailed comments! I know it's painful to review such a large codebase... Highly appreciate your effort!
In this commit, I have addressed most of the comments from Karmel and Kathy on the code details. I have fixed some of Victor's comments, and will work on others mainly about optimization. I also do some clean up/gpylint on the py files, but not all of them. Will try to finish them all soon. Will also add some unit tests on the code.
Feel free to add any comments you may have. Thanks!

yhliang2018 commented:
Thank you, @bitfort ! Still working on it, and will make it better on the following iterations like you said :)
